### PR TITLE
[superfluid] TotalDelegationByValidatorForDenom query

### PR DESCRIPTION
**Pull request by RusAkh**
Closes: #2282

## What is the purpose of the change
Query for getting all sfs'd assets for each validator

## Brief Changelog

added query and test

## Testing and Verifying
This change added tests and can be verified as follows:

unit test

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? yes
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented?  not documented

### CODE REVIEW

File: x/superfluid/keeper/grpc_query.go [hunk 1]
```diff
@@ -392,6 +392,41 @@ func (q Querier) EstimateSuperfluidDelegatedAmountByValidatorDenom(goCtx context
...
+		val, found := q.Keeper.sk.GetValidator(ctx, valAddr)
+		if !found {
+			return nil, stakingtypes.ErrNoValidatorFound
+		}
+		equivalentAmountOSMO := q.Keeper.GetSuperfluidOSMOTokens(ctx, req.Denom, val.Tokens)
+		result := &types.Delegations{
+			ValAddr:        valAddr.String(),
+			AmountSfsd:     &val.Tokens,
+			OsmoEquivalent: &equivalentAmountOSMO,
+		}
```
Comment (mattverse) [inline x/superfluid/keeper/grpc_query.go:420]: hrmm this seems wrong to me. Returning val.Tokens would return all the tokens on the validator, not the amount on the intermediary account. Instead I think we should be doing
```
delegation, found := suite.App.StakingKeeper.GetDelegation(suite.Ctx, expAcc.GetAccAddress(), valAddr)
```

and then be using delegation.shares to track how much is sfsd

File: x/superfluid/keeper/grpc_query_test.go [hunk 2]
```diff
@@ -16,6 +16,58 @@ func (suite *KeeperTestSuite) TestGRPCParams() {
 	suite.Require().True(res.Params.MinimumRiskFactor.Equal(types.DefaultParams().MinimumRiskFactor))
 }
 
+func (suite *KeeperTestSuite) TestTotalDelegationByValidatorForAsset() {
+	suite.SetupTest()
+	ctx := suite.Ctx
+	querier := suite.querier
+	delegation_amount := int64(1000000)
+
+	var denom string
```
Comment (mattverse) [inline x/superfluid/keeper/grpc_query_test.go:25]:
```suggestion

```

**Review by mattverse — COMMENTED** [review]
Hey @RusAkh! Thanks for this PR! I do agree with you that this query is something that we def need!

I think ur on the right track on this PR, just that how sfsd asset amount is calculated seems that it needs further changes to me. Left some comment, please take a look!

**Comment by RusAkh** [comment]
Thanks for the review! I will work on it!

File: x/superfluid/keeper/grpc_query.go (hunk 1 above)
Comment (RusAkh) [reply x/superfluid/keeper/grpc_query.go:420]: @mattverse I do not really understand why do we need to get amount of an intermediary account if we want to get assets of each validator for a specific denomination? I would be very thankful for further clarification, please

Comment (RusAkh) [reply x/superfluid/keeper/grpc_query.go]: other comments were addressed

Comment (mattverse) [reply x/superfluid/keeper/grpc_query.go]: Awesome! Good question, here we want to return the amount superfluid staked, not the amount total staked on the validator right? Say for example valA has total of 100uosmo staked, 50uosmo staked natively, 50uosmo staked via superfluid staking . `val.Tokens` would be returning 100uosmo, whilst using getDelegation would be returning 50uosmo, only the amount that has been staked

Comment (RusAkh) [reply x/superfluid/keeper/grpc_query.go]: oh, that was an awesome example! Thank you! I will try to figure out how to fix that!

Comment (RusAkh) [reply x/superfluid/keeper/grpc_query.go]: @mattverse this comment was also addressed

File: x/superfluid/keeper/grpc_query_test.go [hunk 3]
```diff
@@ -16,6 +16,55 @@ func (suite *KeeperTestSuite) TestGRPCParams() {
...
+		{0, 0, 0, delegation_amount},
+		{0, 1, 1, delegation_amount},
+		{1, 0, 1, delegation_amount},
+		{1, 1, 0, delegation_amount},
+	}
+
+	suite.setupSuperfluidDelegations(valAddrs, superfluidDelegations, denoms)
+
+	for _, denom := range denoms {
+		req, err := querier.TotalDelegationByValidatorForDenom(sdk.WrapSDKContext(ctx), &types.QueryTotalDelegationByValidatorForDenomRequest{Denom: denom})
```
Comment (mattverse) [inline x/superfluid/keeper/grpc_query_test.go:38]:
```suggestion
		res, err := querier.TotalDelegationByValidatorForDenom(sdk.WrapSDKContext(ctx), &types.QueryTotalDelegationByValidatorForDenomRequest{Denom: denom})
```

File: x/superfluid/keeper/grpc_query_test.go [hunk 4]
```diff
@@ -16,6 +16,55 @@ func (suite *KeeperTestSuite) TestGRPCParams() {
...
+
+		suite.Require().NoError(err)
+		suite.Require().Equal(len(valAddrs), len(req.AssetResponse))
```
Comment (mattverse) [inline x/superfluid/keeper/grpc_query_test.go:41]:
```suggestion
		suite.Require().Equal(len(valAddrs), len(res.AssetResponse))
```

File: x/superfluid/keeper/grpc_query_test.go [hunk 5]
```diff
@@ -16,6 +16,55 @@ func (suite *KeeperTestSuite) TestGRPCParams() {
...
+
+		for res_ind, res := range req.AssetResponse {
+			// check osmo equivalent is correct
+			actual_response_osmo := *req.AssetResponse[res_ind].OsmoEquivalent
+			needed_response_osmo := suite.App.SuperfluidKeeper.GetSuperfluidOSMOTokens(ctx, denom, sdk.NewInt(delegation_amount))
+
+			suite.Require().Equal(actual_response_osmo, needed_response_osmo)
+
```
Comment (mattverse) [inline x/superfluid/keeper/grpc_query_test.go:49]: should we also add check for correct osmo equivilent amount in the res?

File: x/superfluid/keeper/grpc_query.go [hunk 6]
```diff
@@ -392,6 +392,46 @@ func (q Querier) EstimateSuperfluidDelegatedAmountByValidatorDenom(goCtx context
 	}, nil
 }
 
+func (q Querier) TotalDelegationByValidatorForDenom(goCtx context.Context, req *types.QueryTotalDelegationByValidatorForDenomRequest) (*types.QueryTotalDelegationByValidatorForDenomResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	var delegationsByValidator = []*types.Delegations{}
```
Comment (mattverse) [inline x/superfluid/keeper/grpc_query.go:398]: So I figured using pointers here is awkward, which made me follow into the rabbit hole until I noticed that the response struct itself is taking pointer referenced array. 

```
type QueryTotalDelegationByValidatorForDenomResponse struct {
	AssetResponse []*Delegations `protobuf:"bytes,1,rep,name=asset_response,json=assetResponse,proto3" json:"asset_response,omitempty"`
}
```

whilst I think it should instead be 

```

type QueryTotalDelegationByValidatorForDenomResponse struct {
	AssetResponse []Delegations `protobuf:"bytes,1,rep,name=asset_response,json=assetResponse,proto3" json:"asset_response,omitempty"`
}
```
just like other proto response structs do. I do think that the possible solution for this is adding `[ (gogoproto.nullable) = false ];` to the proto field. Please lmk if it works out

**Review by mattverse — CHANGES_REQUESTED** [review]
Going through a second round of review, I think we need change in how the proto struct response returns a pointer referenced slice.

File: proto/osmosis/superfluid/query.proto [hunk 7]
```diff
@@ -53,6 +53,10 @@ service Query {
         "/osmosis/superfluid/v1beta1/connected_intermediary_account/{lock_id}";
   }
 
+  rpc TotalDelegationByValidatorForDenom(
```
Comment (alexanderbez) [inline proto/osmosis/superfluid/query.proto:56]: nit: can we add docs for new proto types please 🙏

File: x/superfluid/keeper/grpc_query.go (hunk 6 above)
Comment (alexanderbez) [reply x/superfluid/keeper/grpc_query.go:398]: We should definitely do this ^

**Review by alexanderbez — CHANGES_REQUESTED** [review]

**Comment by RusAkh** [comment]
@alexanderbez @mattverse changes addressed

File: proto/osmosis/superfluid/query.proto [hunk 8]
```diff
@@ -153,6 +158,18 @@ message ConnectedIntermediaryAccountResponse {
...
+  repeated Delegations asset_response = 1;
+}
+
+message Delegations {
+  string val_addr = 1;
+  string amount_sfsd = 2
+      [ (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Int" ];
+  string osmo_equivalent = 3
+      [ (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Int" ];
+}
```
Comment (mattverse) [inline proto/osmosis/superfluid/query.proto:172]:
```suggestion
  string amount_sfsd = 2 [
      (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Int",
      (gogoproto.moretags) = "yaml:\"amount_sfsd\""
    ];
  string osmo_equivalent = 3 [
      (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Int" ],
      (gogoproto.moretags) = "yaml:\"osmo_equivalent\""
 ];  
}
```

File: x/superfluid/keeper/grpc_query.go (hunk 6 above)
Comment (mattverse) [reply x/superfluid/keeper/grpc_query.go:398]: Could we address this before merging? Thanks!

**Comment by RusAkh** [comment]
@mattverse done

File: x/superfluid/keeper/grpc_query.go [hunk 9]
```diff
@@ -392,6 +392,46 @@ func (q Querier) EstimateSuperfluidDelegatedAmountByValidatorDenom(goCtx context
 	}, nil
 }
 
+func (q Querier) TotalDelegationByValidatorForDenom(goCtx context.Context, req *types.QueryTotalDelegationByValidatorForDenomRequest) (*types.QueryTotalDelegationByValidatorForDenomResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	var delegationsByValidator = []types.Delegations{}
+	var intermediaryAccount types.SuperfluidIntermediaryAccount
```
Comment (alexanderbez) [inline x/superfluid/keeper/grpc_query.go:399]: nit: group vars :p

**Review by alexanderbez — APPROVED** [review]

File: proto/osmosis/superfluid/query.proto [hunk 10]
```diff
@@ -153,6 +158,24 @@ message ConnectedIntermediaryAccountResponse {
   SuperfluidIntermediaryAccountInfo account = 1;
 }
 
+message QueryTotalDelegationByValidatorForDenomRequest { string denom = 1; }
+message QueryTotalDelegationByValidatorForDenomResponse {
+  repeated Delegations asset_response = 1 [ (gogoproto.nullable) = false ];
```
Comment (mattverse) [inline proto/osmosis/superfluid/query.proto:163]:
```suggestion
  repeated Delegations assets = 1 [ (gogoproto.nullable) = false ];
```

**Review by mattverse — APPROVED** [review]
LGTM! Nice work!

**Pull request merged**